calendarプログラムを作成

sthnu32424 opened the pull request.

sthnu32424 committed:
Add prototype of calendar

sthnu32424 committed:
Rename variables and clean up default values

- Renamed variables for better readability
- Removed unnecessary `.to_int` calls
- Simplified default year/month handling for better clarity
- Replaced `params` keys (`:m` → `:month`, `:y` → `:year`) for consistency

sthnu32424 committed:
Refactor code by extracting methods for better readability

sthnu32424 committed:
Refactor indentation and enhance calendar header formatting

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,38 @@
+  puts calendar_rows
+end
+
+parse_arguments
```
`parse_arguments`を呼ぶことでカレンダー表示まで全て行われる、というのは読みにくいです。以下リンク先のような話と近いかなと思いますが、いかがですか? https://bootcamp.fjord.jp/pages/avoid-matroska-design `parse_arguments`というなら`parse_arguments`だけを仕事にして欲しいかなーと。

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,38 @@
+  option_parser.on('-y [VAL]', Integer) { |m| params[:year] = m }
+  option_parser.parse!(ARGV, into: params)
+  params[:year] ||= Date.today.year
+  params[:month] ||= (params[:year] == Date.today.year ? Date.today.month : 1)
```
`Date.today`が何度も呼ばれてますが、厳密には一度目と二度目の呼び出しタイミングでちょうど年跨ぎになったりすると動作がおかしくなるのではないですかね。実行時の年月は一度だけにして固定しておくほうが安全なプログラムになるのではないかなと思います。今回くらいのプログラムではまず問題にならないと思いますが、例えば会計処理の締め処理等でこういったところがあると、ある日突然原因のよくわからない不具合が起きたりしそうです。

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,38 @@
+
+def generate_calendar(params)
+  first_day = Date.new(params[:year], params[:month], 1)
+  last_day = Date.new(params[:year], params[:month], -1)
```
細かいことを言うと`Date`型ですし、`xxx_date`という命名がいいかもしれませんね。`last_day.day`のように同じ単語が重なるような呼び出しも防げますし。`date`だと年月日、`day`だと単に日を表すようなものだと理解してます。(英語得意ではないので違うかも...?)

sthnu32424 committed:
refactor: avoid multiple calls to Date.today

when the year is crossed, the operation goes wrong
https://github.com/shou-317/ruby-practices/pull/2#discussion_r1947476460

sthnu32424 committed:
refactor: change first_day and last_day to first_date and last_date

change to the format 'xxx_date' because it is a date type
https://github.com/shou-317/ruby-practices/pull/2#discussion_r1947477935

sthnu32424 committed:
refactor: clarify the order of execution with main

because the calendar is even displayed in parse arguments
https://github.com/shou-317/ruby-practices/pull/2#discussion_r1947475637

sthnu32424 committed:
refactor: remove unnecessary assignments to calendar_rows

because the change in processing order causes a warning in rubocop
